assets/src/vo/date-time/datetime.js

```diff
@@ -0,0 +1,1156 @@
+/**
+ * External imports
+ */
+import moment from 'moment-timezone';
+import {
+	capitalize,
+	omit,
+	isNumber,
+	isEmpty,
+	reduce,
+	isObject,
+	isUndefined,
+} from 'lodash';
+
+/**
+ * Internal imports
+ */
+import {
+	InvalidDateTime,
+	InvalidArgument,
+} from '@eventespresso/eejs';
+import * as assertions from './assertions';
+import Duration from './duration';
+import {
+	DEFAULT_TIMEZONE_STRING,
+	DEFAULT_OFFSET,
+	DEFAULT_VALID_LOCALE,
+	DEFAULT_FORMAT,
+} from './defaults';
+
+/**
+ * A collection of symbols used for "private" properties in the DateTime object.
+ *
+ * @type {
+ * 	{
+ * 		datetime: Symbol
+ * 	}
+ * }
+ */
+const privateProperties = {
+	datetime: Symbol( 'DateTimePropertyDateTime' ),
+};
+
+/**
+ * A collection of symbols used for "private" methods in the DateTime object.
+ *
+ * @type {
+ * {
+ * 	getUnitNames: Symbol,
+ * 	createGettersAndSetters: Symbol,
+ * 	extractMomentsFromDateTimes: Symbol,
+ * 	normalizeUnitName: Symbol,
+ * 	normalizeUnitObject: Symbol,
+ * 	normalizeUnitValue: Symbol,
+ * 	}
+ * }
+ */
+const privateMethods = {
+	getUnitNames: Symbol( 'DateTimeMethodGetUnitNames' ),
+	createGettersAndSetters: Symbol( 'DateTimeMethodCreateGettersAndSetters' ),
+	extractMomentsFromDateTimes: Symbol( 'DateTimeMethodExtractMomentsFromDateTimes' ),
+	normalizeUnitName: Symbol( 'DateTimeMethodNormalizeUnitName' ),
+	normalizeUnitObject: Symbol( 'DateTimeMethodNormalizeUnitObject' ),
+	normalizeUnitValue: Symbol( 'DateTimeMethodNormalizeUnitValue' ),
+};
+
+const validDateTimeUnits = [
+	'year',
+	'month',
+	'day',
+	'hour',
+	'minute',
+	'second',
+	'millisecond',
+];
+
+/**
+ * The DateTime value object represents a single point in time.
+ *
+ * Internally, the DateTime class here uses `moment`.  This is an abstraction
+ * loosely following the adapter pattern so that there is a common api that
+ * can be depended on if in the future the internal library is switched to
+ * something different (such as Luxon).
+ */
+export default class DateTime {
+	/**
+	 * The constructor for the DateTime class
+	 *
+	 * @param {Moment|string} iso8601DateString  The public api
+	 * expects an ISO 8601 date-time string.  However for internal use, a moment
+	 * object is handled as well (for performance reasons).  Client code should
+	 * not construct with a moment object because that will be fragile in the
+	 * case the internal library is changed in the future.
+	 * @param {string} timezone
+	 * @param {string} locale
+	 */
+	constructor(
+		iso8601DateString = '',
+		timezone = DEFAULT_TIMEZONE_STRING,
+		locale = DEFAULT_VALID_LOCALE
+	) {
+		/**
+		 * If a moment object is provided just construct from that.  This will
+		 * not be promoted as public api though.  It's mostly just used
+		 * internally for performance reasons.
+		 */
+		if ( moment.isMoment( iso8601DateString ) ) {
```

Comment:
> What about situations where the incoming parameters are inconsistent with themselves, such as iso8601DateString being a moment object with no timezone information (ie: set to UTC) and timezone being a valid non-UTC timezone?

The js docs for the constructor (and the documentation in docs) state **"The public api expects an ISO 8601 date-time string.  However, for internal use..."** So basically, the only reason why the constructor accepts a moment object is for internal use only.  If developers choose to provide a moment object then it will be expected that they understand (or learn) the ramifications of doing so.

> In other words, should we apply any incoming timezone parameters to iso8601DateString parameters that are moment objects, and/or verify that the supplied moment object is already using the supplied timezone?

I didn't do this by choice because timezone and locale getters always retrieve from the moment object anyways.  The assumption is that if someone is providing a moment object, they've already set that to whatever timezone/locale they want.  Again, via js docs, developers are _discouraged_ from providing a moment object in the constructor.  If someone is starting with a moment object, it should be trivial to do `moment.toISOString()` to pass that in along with timezone and offset stuff (or just pass in the moment object that already has timezone and offset set on it.

> So basically, what I'm saying is that it may be better to process the incoming parameters separately in an incremental fashion as opposed to using an if-else tree.

The problem with your if-else tree example is that one of the primary reasons why a moment object is accepted to begin with is to avoid doing additional assertions/etc on the assumption that the moment object has that information.  The public api _discourages_ providing a moment object because any client code doing that explicitly will break if we ever switch to Luxon.

> Another possibly better solution would be to add some more static factory methods for creating a DateTime from a moment object that has not been fully populated with localization data. For example:

One of the purposes of doing the `Datetime` and `Duration` value objects was to hide the implementation details of what library is used under the hood.  With that said, I do see there being some value in having a `fromMoment` (you wouldn't need `withTimezone`, or `withLocale` variants because that's all available on the moment object), because even if we shift to using `Luxon` (or some other library) down the road, we could just have that method construct from the moment.

I think if anything what I need to do is revisit this so that the constructor does NOT receive a moment object at all.  Internally, we'll just have to refactor things so that instead of passing along a moment object, we pass along the properties extracted from it.  It might not be as performant, but it would make the api much clearer.  Thoughts?

Reply:
As an example, basically `DateTime.fromMoment` would have something like this logic:

```js
static fromMoment( momentObj ) {
	if ( ! moment.isMoment( momentObj ) ) {
		throw new TypeError( 'Not a moment object.' );
    }
	return isFunction( momentObj.tz ) ?
		new DateTime( momentObj.toISOString(), momentObj.tz(), momentObj.locale() ) :
		new DateTime( momentObj.toISOString( true ), 'UTC', momentObj.locale() );
}
```

Then internally instead of passing moment to the constructor for when we have a moment object, we just call `DateTime.fromMoment()`.